### PR TITLE
#81 -- add tests to confirm exception handling in views

### PR DESCRIPTION
**Pull request by monster-michael**
Not sure if this is what you are looking for, but getting to know mock was fun either way. I tried to stick with existing code conventions. 

Let me know if you have any suggestions.

### CODE REVIEW

**Comment by monster-michael** [comment]
Looks like that all failed. Looking into it now.

File: tests/callistocore/test_views.py [hunk 1]
```diff
@@ -787,6 +810,36 @@ def test_match_sends_custom_report(self):
         self.assertIn('test match delivery body', message.body)
         self.assertRegexpMatches(message.attachments[0][0], 'custom_.*\\.pdf\\.gpg')
 
+    @patch('callisto.delivery.views.MatchReport.encrypt_match_report')
+    def test_match_send_exception_is_handled(self, mock_encrypt_match_report):
+        mock_encrypt_match_report.side_effect = Exception('Mock Submit Match Exception')
```
Comment (kelseyq) [inline tests/callistocore/test_views.py]: mocks look great!

File: callisto/delivery/views.py [hunk 2]
```diff
@@ -111,6 +111,7 @@ def submit_to_school(request, report_id, form_template_name="submit_to_school.ht
             except Exception:
                 # report was sent even if confirmation email fails, so don't show an error if so
                 logger.exception("couldn't send confirmation to user on submission")
+                context.update({'form': form, 'email_confirmation_error': True})
```
Comment (kelseyq) [inline callisto/delivery/views.py]: I'd rather not add extra context to the response solely for tests. in the test, can we confirm that two things have happened: 1) the report was sent to school (best way to do this is probably with another mock) and 2) the confirmation template is the template that's returned?

File: callisto/delivery/views.py [hunk 3]
```diff
@@ -176,6 +177,7 @@ def submit_to_matching(request, report_id, form_template_name="submit_to_matchin
             except Exception:
                 # matching was entered even if confirmation email fails, so don't show an error if so
                 logger.exception("couldn't send confirmation to user on match submission")
+                context.update({'email_confirmation_error': True})
```
Comment (kelseyq) [inline callisto/delivery/views.py]: same as above

File: callisto/delivery/views.py [hunk 4]
```diff
@@ -224,6 +226,7 @@ def export_as_pdf(request, report_id, force_download=True, filename='report.pdf'
             except Exception:
                 logger.exception("could not export report {}".format(report_id))
                 form.add_error(None, "There was an error exporting your report.")
+                context.update({'generate_pdf_error': True})
```
Comment (kelseyq) [inline callisto/delivery/views.py]: for these, you should be able to get the form from the context and verify that the correct error was added to it in line 228

File: callisto/delivery/views.py (hunk 2 above)
Comment (monster-michael) [inline callisto/delivery/views.py]: Agreed. I felt weird about changing the views.py so this is better. My only concern with checking if the report was actually sent is that you will most likely be sending emails async at some point in time and the delay may not be consistent - so tracking emails are actually gone doesn't seem to match a potential real world scenario.

File: tests/callistocore/test_views.py [hunk 5]
```diff
@@ -787,6 +813,39 @@ def test_match_sends_custom_report(self):
...
+                                    data={'name': 'test submitter',
+                                          'email': 'test@example.com',
+                                          'phone_number': '555-555-1212',
+                                          'email_confirmation': "True",
+                                          'key': self.report_key,
+                                          'form-0-perp': 'facebook.com/test_url',
+                                          'form-TOTAL_FORMS': '1',
+                                          'form-INITIAL_FORMS': '1',
+                                          'form-MAX_NUM_FORMS': '', })
+        self.assertTrue(mock_logger.exception.called)
```
Comment (kelseyq) [inline tests/callistocore/test_views.py]: hey Michael--good call thinking about future async emails. I think mocking the call that queues an email to be sent/makes the async call would still work, bu this is a great solution too.

**Pull request merged**